### PR TITLE
import deployment package in jobs.go to register the deployment webhook

### LINKED ISSUE #2987: deployment webhook is broken
**What happened**:
1. Installed kueue using the following commands:
```
IMAGE_REGISTRY=<registry> make  image-local-push deploy
```
2. Ran patch command to try out the visibility api
```
 kubectl patch -n kueue-system deployment kueue-controller-manager --type='json' -p='[

  {
    "op": "add",
    "path": "/spec/template/spec/containers/0/args/-",
    "value": "--feature-gates=VisibilityOnDemand=true"
  }
]'
```

The error failed with 
```
Error from server (InternalError): Internal error occurred: failed calling webhook "vdeployment.kb.io": failed to call webhook: the server could not find the requested resource
```
3. Tried to create busybox deployment, failed with the same error

**What you expected to happen**:
Create and update calls to deployment should work.

**How to reproduce it (as minimally and precisely as possible)**:
1. checkout main
2. run `IMAGE_REGISTRY=<registry> make  image-local-push deploy`
3. create or update deployment as noted above

**Anything else we need to know?**:
This PR #2813 added a validating and mutating webhook for deployments, but it is not registered by default

**Environment**:
- Kubernetes version (use `kubectl version`): tested in 1.30.3 and 1.31.0
- Kueue version (use `git describe --tags --dirty --always`): main at commit  0bcde8c82e6a769bf75e3d195b65d18dbb6e2bf6
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by alaypatel07**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This PR fixes deployment webhook to allow for create/update of deployments after kueue is installed

#### Which issue(s) this PR fixes:

Fixes #2987

#### Special notes for your reviewer:
Testing done:
1. enable the pod controller with following configuration option and applied the patch
```
$ cat config/components/manager/controller_manager_config.yaml
...
  - "pod"
  - "deployment"
#  externalFrameworks:
#  - "Foo.v1.example.com"
  podOptions:
    namespaceSelector:
      matchExpressions:
        - key: kubernetes.io/metadata.name
          operator: NotIn
          values: [ kube-system, kueue-system ]
...
$  kubectl patch -n kueue-system deployment kueue-controller-manager --type='json' -p='[
  {
    "op": "add",
    "path": "/spec/template/spec/containers/0/args/-",
    "value": "--feature-gates=VisibilityOnDemand=true"
  }
]'
deployment.apps/kueue-controller-manager patched
```
2. disabled the pod controller and applied the patch
```
$ cat config/components/manager/controller_manager_config.yaml
...
#  - "pod"
#  externalFrameworks:
#  - "Foo.v1.example.com"
#  podOptions:
#    namespaceSelector:
#      matchExpressions:
#        - key: kubernetes.io/metadata.name
#          operator: NotIn
#          values: [ kube-system, kueue-system ]
...
$ kubectl patch -n kueue-system deployment kueue-controller-manager --type='json' -p='[
  {
    "op": "add",
    "path": "/spec/template/spec/containers/0/args/-",
    "value": "--feature-gates=VisibilityOnDemand=true"
  }
]'
deployment.apps/kueue-controller-manager patched
```

Both the scenarios worked

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by alaypatel07** [comment]
cc @mbobrovskyi since you were interested in the issue, PTAL

File: config/components/webhook/manifests.yaml [hunk 1]
```diff
@@ -4,6 +4,25 @@ kind: MutatingWebhookConfiguration
 metadata:
   name: mutating-webhook-configuration
 webhooks:
+- admissionReviewVersions:
```
Comment (mimowo) [inline config/components/webhook/manifests.yaml:7]: Is this change required, or just formatting? If formatting, then I would suggest to revert.

File: pkg/controller/jobframework/integrationmanager.go [hunk 2]
```diff
@@ -265,7 +267,13 @@ func GetIntegration(name string) (IntegrationCallbacks, bool) {
 func GetIntegrationByGVK(gvk schema.GroupVersionKind) (IntegrationCallbacks, bool) {
 	for _, name := range manager.getList() {
 		integration, ok := GetIntegration(name)
-		if ok && gvk == integration.NewJob().GVK() {
+		var g GVKer
```
Comment (mimowo) [inline pkg/controller/jobframework/integrationmanager.go:270]: Can you explain why this is needed? Also, this check could be done only when `ok` - so you could do `if ok && matchingGVK(integration, gvk)`

File: pkg/controller/jobframework/integrationmanager.go [hunk 3]
```diff
@@ -265,7 +267,13 @@ func GetIntegration(name string) (IntegrationCallbacks, bool) {
...
+		if integration.NewJob != nil {
+			g = integration.NewJob()
+		} else {
+			g = integration.GVKer()
+		}
```
Comment (mbobrovskyi) [inline pkg/controller/jobframework/integrationmanager.go:275]: Can we add `NewJob` callback on `deployment_controller` instead to create another one?

File: config/components/webhook/manifests.yaml (hunk 1 above)
Comment (mbobrovskyi) [reply config/components/webhook/manifests.yaml:7]: I believe it's automatically added on `make verify`.

Comment (mimowo) [reply config/components/webhook/manifests.yaml:7]: Hm, but it is only moved from another place. Why wasn't this then moved prior to this PR?

Comment (mbobrovskyi) [reply config/components/webhook/manifests.yaml:7]: I think it's because `_ "sigs.k8s.io/kueue/pkg/controller/jobs/deployment"` added  on `pkg/controller/jobs/jobs.go` and sort list changed on code-generator. In any case, `make verify` should throw an error if something not correct.

Comment (mimowo) [reply config/components/webhook/manifests.yaml:7]: sgtm, thanks!

File: pkg/controller/jobframework/integrationmanager.go (hunk 3 above)
Comment (alaypatel07) [reply pkg/controller/jobframework/integrationmanager.go:275]: > NewJob and jobframework.GenericJob would require a lot of methods to be implemented, probably that's why

This is correct, I started out implementing the interface for deployment_controller, but a lot of the methods there are not applicable because it being no-op controller.  For all no-op controllers, there is a need to find a way to avoid implementing GenericJob interface.

Comment (mbobrovskyi) [reply pkg/controller/jobframework/integrationmanager.go:275]: Can we implement jobframework.GenericJob? If I correct understood it's required option https://kueue.sigs.k8s.io/docs/tasks/dev/integrate_a_custom_job/#job-framework no?

File: pkg/controller/jobframework/integrationmanager.go (hunk 2 above)
Comment (alaypatel07) [reply pkg/controller/jobframework/integrationmanager.go:270]: >  Also, this check could be done only when ok - so you could do if ok && matchingGVK(integration, gvk)

The check on ok passes, but the ingetration.NewJob() is nil for deployment_controller. And like @vladikkuzn said setting NewJob to a non-nil value means implementing GenericJob interface for a deployment_controller which didn't make sense. The idea was that GenericJob is a big interface, decomposing it into smaller ones will help with other no-op controllers like deployment_controller in the future.

Comment (mimowo) [reply pkg/controller/jobframework/integrationmanager.go:270]: Thanks for explaining, sgtm. In the second part of the comment I meant just a small refactor:

```golang
func matchingGVK(integration, gvk) bool {
  if integration.NewJob != nil {
      return gvk == integration.NewJob().GVK()
   } else {
    return gvk == integration.GVKer().GVK()
  }
}
```
then use it `if ok && gvk == matchingGVK(integration, gvk) {`, to avoid calling NewJob if `!ok`. Would this work?

File: pkg/controller/jobframework/integrationmanager.go (hunk 3 above)
Comment (alaypatel07) [reply pkg/controller/jobframework/integrationmanager.go:275]: > WDYT about getting gvk from JobType?

This seems reasonable. Given that we can find GVK from JobType, I will go ahead and remove that method form the interface?

Comment (mbobrovskyi) [reply pkg/controller/jobframework/integrationmanager.go:275]: The problematic part that we are using GenericJob on the kueuectl to filter pods by Job https://github.com/kubernetes-sigs/kueue/blob/ece18974690b4e59cfa179b4514da0cffcd53fda/cmd/kueuectl/app/list/list_pods.go#L231-L240.

Comment (mbobrovskyi) [reply pkg/controller/jobframework/integrationmanager.go:275]: > > Can we implement jobframework.GenericJob? If I correct understood it's required option https://kueue.sigs.k8s.io/docs/tasks/dev/integrate_a_custom_job/#job-framework no?
> 
> The problem that it's not a job :) It's "serving workload" I don't think it can be suspended and so on

If it is not a Job should we move it to another directory?

Comment (mimowo) [reply pkg/controller/jobframework/integrationmanager.go:275]: > If it is not a Job should we move it to another directory?

This might be a good idea, like "workload", but I would suggest to focus this PR on fixing the issue, and we could follow up with a structure cleanup.

Comment (alaypatel07) [reply pkg/controller/jobframework/integrationmanager.go:275]: @vladikkuzn checking from jobtype is not working
> WDYT about getting gvk from JobType?

Reason being the JobType objects are empty, for example
```
		JobType:        &appsv1.Deployment{},
``` 
in deployment controller hence their gvks are empty, what we need is a filled GVK like 
```
    gvk = appsv1.SchemeGroupVersion.WithKind("Deployment")
```

Comment (alaypatel07) [reply pkg/controller/jobframework/integrationmanager.go:275]: > The problematic part that we are using GenericJob on the kueuectl to filter pods by Job

sorry I am not following what the problem is here? IIUC the deployment controller has to be in jobframework because of its dependencies on the pod controller which is an actual implementation of GenericJob

File: pkg/controller/jobframework/integrationmanager.go [hunk 4]
```diff
@@ -56,6 +56,8 @@ type ReconcilerFactory func(client client.Client, record record.EventRecorder, o
 type IntegrationCallbacks struct {
 	// NewJob creates a new instance of job
 	NewJob func() GenericJob
+	// GVKer returns the GVK for job
+	GVKer func() GVKer
```
Comment (alculquicondor) [inline pkg/controller/jobframework/integrationmanager.go:60]:
```suggestion
	GVK schema.GroupVersionKind
```

Comment (alculquicondor) [reply pkg/controller/jobframework/integrationmanager.go:60]: It doesn't look like we need the indirection.

**Comment by alaypatel07** [comment]
Considering all the input, I have reworked the PR, PTAL

File: cmd/kueuectl/app/list/list_pods.go [hunk 5]
```diff
@@ -228,6 +228,9 @@ func (o *PodOptions) getPodLabelSelector() (string, error) {
 		return "", nil
 	}
 
+	if cbs.NewJob == nil {
+		return "", fmt.Errorf("NewJob is not defined for integration job")
```
Comment (mbobrovskyi) [inline cmd/kueuectl/app/list/list_pods.go:232]:
```suggestion
		return "", nil
```

We already have an error [here](https://github.com/kubernetes-sigs/kueue/blob/c7b367ec851b862f831fa5df955598702f0116ee/cmd/kueuectl/app/list/list_pods.go#L107).

File: pkg/controller/jobframework/integrationmanager.go [hunk 6]
```diff
@@ -56,6 +56,8 @@ type ReconcilerFactory func(client client.Client, record record.EventRecorder, o
 type IntegrationCallbacks struct {
 	// NewJob creates a new instance of job
 	NewJob func() GenericJob
+	// GVK holds the schema information for the job
```
Comment (mbobrovskyi) [inline pkg/controller/jobframework/integrationmanager.go:59]:
```suggestion
	// GVK holds the schema information for the job
	// (this callback is optional)
```

Comment (mbobrovskyi) [reply pkg/controller/jobframework/integrationmanager.go:59]: Or maybe we can implement GVKs in all jobs. @alculquicondor WDYT?

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

File: pkg/controller/jobframework/integrationmanager.go [hunk 7]
```diff
@@ -265,7 +267,10 @@ func GetIntegration(name string) (IntegrationCallbacks, bool) {
 func GetIntegrationByGVK(gvk schema.GroupVersionKind) (IntegrationCallbacks, bool) {
 	for _, name := range manager.getList() {
 		integration, ok := GetIntegration(name)
-		if ok && gvk == integration.NewJob().GVK() {
+		if ok && integration.NewJob == nil && integration.GVK == gvk {
+			return integration, true
+		}
+		if ok && integration.NewJob != nil && integration.NewJob().GVK() == gvk {
 			return integration, true
 		}
```
Comment (mbobrovskyi) [inline pkg/controller/jobframework/integrationmanager.go:275]: I would like to use @mimowo [suggestion](https://github.com/kubernetes-sigs/kueue/pull/2988#discussion_r1745789652).

```
func matchingGVK(integration, gvk) bool {
  if integration.NewJob != nil {
      return gvk == integration.NewJob().GVK()
   } else {
    return gvk == integration.GVK
  }
}
```

WDYT?

File: pkg/controller/jobframework/integrationmanager.go (hunk 6 above)
Comment (alaypatel07) [reply pkg/controller/jobframework/integrationmanager.go:59]: Can we defer the implementation changes across the board until we find a better way? as noted here https://github.com/kubernetes-sigs/kueue/pull/2988#discussion_r1745809416 

I will make the comment update. If someone can help me create an issue for the bigger refactor, I can start looking at it

Comment (alculquicondor) [reply pkg/controller/jobframework/integrationmanager.go:59]: We could remove the `GVK` method from the Job interface and just populate the IntegrationCallbacks struct. We have a gvk constant for every integration already, IIRC.

But I'm ok leaving it for another PR.

File: pkg/controller/jobframework/integrationmanager.go (hunk 7 above)
Comment (alaypatel07) [reply pkg/controller/jobframework/integrationmanager.go]: yes, missed that, updated now.

**Comment by mbobrovskyi** [comment]
Thanks!

/lgtm

**Comment by alaypatel07** [comment]
@alculquicondor can you please take another look and give out the approve tag? TIA
